news/2025-03-27-spectec/index.md

```diff
@@ -0,0 +1,218 @@
+the picture), that the _prose specification_ that you'll read in the rendered
+document actually is correct and defines what we think it defines! That was not
+the case before, and indeed, while developing SpecTec, we discovered numerous
+bugs in the hand-crafted prose, and there are probably many more we didn't spot.
+With SpecTec, a wide range of brainless spec bugs becomes impossible, some by
+construction, some because we can run it against the test suite. (Caveat: The
+meta interpreter currently calls out to the Wasm reference interpreter for
+decoding and validation; hence this only tells us something about execution. In
+the future, we hope to make it more self-contained.)
+
+In a similar vein, SpecTec can generate definitions representing the Wasm
+semantics in [Coq](https://coq.inria.fr), the dominant proof assistant in the
+area of programming languages. We have started porting the proofs from the
+hand-written WasmCert mechanization, and can already handle soundness for the
+Wasm 1.0 subset. Scaling to the full language is work in progress.
+
+Once we have proved full soundness this way, we also reach a new level of
+assurance regarding the _formal specification_ that you see in the rendered
+specification document! This is because both rendered spec and mechanized
+definitions are now auto-generated from the same source of truth, and producing
+Coq no longer involves a manual translation that is merely based on "eyeball
+correspondence" with the original paper rules.
+
+Finally, we have recently started work on another exciting backend for SpecTec:
+this one can spit out `.wast` test files. So, it is essentially a test fuzzer.
+However, it is a fuzzer that is guided by full knowledge of the syntax and
+semantics of the language! It understands the typing rules and can predict the
+result of executing a piece of code. For example, by just looking at the typing
+rule of an instruction, it could systematically generate a complete
+combinatorial matrix of micro tests for that instruction, that is, tests for its
+behavior under all interesting combinations of types and operands.
+
+Once more, this would take Wasm to a new level of assurance, this time for
+implementations. So far, it has been tedious manual work to write tests — or in
+some rare cases, manually written scripts that generate some very specific
+tests. The coverage of the Wasm test suite, despite not being bad overall, hence
+has been "variable" at best. In the future, we hope to get much better coverage
+with much less work. And because SpecTec is a generic tool, this ought to be
+applicable not just to present features, but also to features that Wasm grows in
+the future.
+
+As a final remark, note that SpecTec is _not_ AI. Instead, it is a meticulously
+designed translation process. That is very important: when accuracy and rigor is
+the goal, then AI with its blackbox behavior and tendency to hallucinate is not
+an adequate tool.
```

Comment:
I fully agree with this, but I don't think editorializing about AI fits well with the rest of the article, especially given the forum where it is being published and the variety of interests represented by the Wasm CG.